internal/controller/ovnk_ipfix_collector.go

```diff
@@ -0,0 +1,291 @@
+	"errors"
+	"net"
+	"os"
+	"time"
+
+	pb "github.com/illumio/cloud-operator/api/illumio/cloud/k8sclustersync/v1"
+	netflows "github.com/netsampler/goflow2/decoders/netflow"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OVNKFlow represents a flow captured from OVN-Kubernetes.
+type OVNKFlow struct {
+	SourceIP        string
+	DestinationIP   string
+	SourcePort      uint16
+	DestinationPort uint16
+	Protocol        string
+	IPVersion       string
+	StartTimestamp  *timestamppb.Timestamp
+}
+
+// https://datatracker.ietf.org/doc/rfc7011/
+// The IPFIX Message Header 16-bit Length field limits the length of an
+// IPFIX Message to 65535 octets, including the header.  A Collecting
+// Process MUST be able to handle IPFIX Message lengths of up to
+// 65535 octets.
+const ipfixMaxLength = 65535
+
+// isOVNKDeployed checks for the presence of the `openshift-ovn-kubernetes` namespace, detection based on OVN-K namespace, this is configurable.
+// https://ovn-kubernetes.io/installation/launching-ovn-kubernetes-on-kind/#run-the-kind-deployment-with-podman
+func (sm *streamManager) isOVNKDeployed(logger *zap.Logger, ovnkNamespace string) bool {
+	logger.Debug("Checking for OVN-K deployment")
+	clientset, err := NewClientSet()
+	if err != nil {
+		logger.Error("Failed to create clientset", zap.Error(err))
+		return false
+	}
+	_, err = clientset.CoreV1().Namespaces().Get(context.TODO(), ovnkNamespace, metav1.GetOptions{})
+	if err != nil {
+		logger.Debug("OVN-K namespace not found", zap.Error(err))
+		return false
+	}
+	logger.Debug("OVN-K namespace found")
+	return true
+}
+
+// startOVNKIPFIXCollector starts the OVN-K IPFIX collector.
+// It creates a custom listener for OVN-K flows on port 4739 and processes
+// the received flows in a separate goroutine.
+func (sm *streamManager) startOVNKIPFIXCollector(ctx context.Context, logger *zap.Logger) error {
+	logger.Info("Starting OVN-K IPFIX Collector")
+	ctxProcessOVNKFlow, cancelProcessOVNKFlow := context.WithCancel(ctx)
+	defer cancelProcessOVNKFlow()
+	go func() {
+		logger.Debug("Starting processIPFIXMessage goroutine")
+		if err := sm.processIPFIXMessage(ctxProcessOVNKFlow, logger); err != nil {
+			logger.Error("Error in processIPFIXMessage", zap.Error(err))
+		}
+	}()
+	select {
+	case <-ctx.Done():
+		logger.Info("Context canceled in startOVNKIPFIXCollector")
+		return ctx.Err()
+	default:
+		logger.Info("Creating custom listener for OVN-K flows")
+		listener, err := net.ListenPacket("udp", ":4739")
+		if err != nil {
+			logger.Fatal("Failed to listen on OVN-K port", zap.Error(err))
+		}
+
+		logger.Info("OVN-k server listening", zap.String("address", sm.streamClient.IPFIXCollectorPort))
+		for {
+			buf := make([]byte, ipfixMaxLength)
+			select {
+			case <-ctx.Done():
+				logger.Info("Context canceled in OVN-K listener loop")
+				return ctx.Err()
+			default:
+				n, _, err := listener.ReadFrom(buf)
```

Comment:
This can block indefinitely, until the listener is closed.
The problem is that nothing in this code is closing the listener.
The listener should be closed when the context is canceled.
So the only proper way to do this is to do these `ReadFrom` blocking calls in a separate goroutine, and have this main function here wait for the context to be canceled and then close the listener, which will unblock any `ReadFrom` blocking call in the goroutine.`